Add announcement for 2.0.0.alpha5 (#537)
hanami/hanami.github.io: source/blog/2022-01-12-announcing-hanami-200alpha5.html.markdown

---
title: Announcing Hanami v2.0.0.alpha5
date: 2022-01-12 11:00:00 UTC
tags: announcements
author: Tim Riley
image: true
excerpt: >

>   Sensible default logger, full source dirs config, router loading flexibility, route helpers in views

Reviewer: ```suggestion
  Sensible default logger, full source dirs config, faster router loading for code reloading, route helpers in views
```

Author: @jodosha Given that we don't have the code reloading side of this sorted out yet, I think I might not apply this suggestion of yours. I reckon it's better for us to tell that part of the story when we have the working code to back it up.

---

Happy new year, Hanami community! To get 2022 started, we're excited to announce the release of Hanami 2.0.0.alpha5!

This release brings the last month of our work on Hanami 2.0 (with an extra week added for good measure, while we all returned from our end of year breaks). It includes:

- Sensible default configuration for the application logger

> - Comprehensive `source_dirs` configuration

Reviewer: ```suggestion
- Comprehensive `config.source_dirs` configuration (for advanced users)
```

Author: 👍🏼 

@jodosha FWIW I've changed this to just "source dirs" instead of `source_dirs`, to make it a more natural-reading title.

> - Lazy router and Rack app initialization

Reviewer: ```suggestion
- Lazy router and Rack app initialization (faster dev reloading)
```

Author: @jodosha n.b. this won't give us faster dev reloading _yet_ — there's still a missing piece of the puzzle, which is to have our app server started non-booted (and properly reloaded upon code changes) in dev-mode only — we don't have those yet.

What I might do is make a note that this opens up this possibility for us, and that we'll be making sure it's in place before 2.0.0 final.

```ruby
module MyApp
  class Application < Hanami::Application
    config.logger.level = :info

    config.logger.stream = $stdout
    config.logger.stream = "/path/to/file"
    config.logger.stream = StringIO.new

    config.logger.format = :json
    config.logger.format = MyCustomFormatter.new

    config.logger.color = false # disable coloring
    config.logger.color = MyCustomColorizer.new

    config.logger.filters << "secret" # add
    config.logger.filters += ["yet", "another"] # add
    config.logger.filters = ["foo"] # replace

    # See https://ruby-doc.org/stdlib/libdoc/logger/rdoc/Logger.html
    config.logger.options = ["daily"] # time based log rotation
    config.logger.options = [0, 1048576] # size based log rotation
  end
end
```

You can also customise the config specifically for given environments:

```ruby
module MyApp
  class Application < Hanami::Application
    config.environment(:staging) do
      config.logger.level = :info
    end
  end
end
```

And finally, you can assign a completely custom logger object:

```ruby
module MyApp
  class Application < Hanami::Application
    config.logger = MyCustomLogger.new
  end
end
```

> ## Comprehensive `source_dirs` configuration

Reviewer: ```suggestion
## Comprehensive `config.source_dirs` configuration (for advanced users)
```

```ruby
module MyApp
  class Application < Hanami::Application
    # Adding a simple component dir
    config.source_dirs.component_dirs.add "serializers"

    # Adding a component dir with custom configuration
    config.source_dirs.component_dirs.add "serializers" do |dir|
      dir.auto_register = proc { |component|
        !component.identifier.start_with?("structs")
      }
    end
  end
end
```

You can also customise the configuration of the default component dirs ("lib", "actions", "repositories", "views"):

```ruby
module MyApp
  class Application < Hanami::Application
    # Customising a default component dir
    config.source_dirs.component_dirs.dir("lib").auto_register = proc { |component|
      !component.identifier.start_with?("structs")
    }

    # Setting default config to apply to all component dirs
    config.source_dirs.component_dirs.auto_register = proc { |component|
      !component.identifier.start_with?("entities")
    }

    # Removing a default component dir
    config.source_dirs.component_dirs.delete("views")
  end
end
```

> The `component_dirs` setting is provided by [dry-system’s](http://dry-rb.org/gems/dry-system/0.21/) own `Dry::System::Config::ComponentDirs`, so you can use this to configure every possible aspect of component loading as you require.

Reviewer: ```suggestion
The `config.component_dirs` setting is provided by [dry-system’s](http://dry-rb.org/gems/dry-system/0.21/) own `Dry::System::Config::ComponentDirs`, so you can use this to configure every possible aspect of component loading as you require.
```

In addition to component dirs, you can also configure your own autoload paths, for source files you don’t want registered as components, but whose classes you still want to access inside your application. These directories are helpful for any classes that you will directly instantiate with their own data rather than dependencies, such as entities, structs, or any other kind of value class.

Out of the box, our `autoload_paths` is `["entities"]`. You can configure this as required:

```ruby
module MyApp
  class Application < Hanami::Application
    # Adding your own autoload paths
    config.source_dirs.autoload_paths << "structs"

    # Or providing a full replacement
    config.source_dirs.autoload_paths = ["structs"]
  end
end
```

## Lazy router and Rack app initialization

> One of the most powerful features of Hanami 2 is your ability to partially boot the application, loading only the specific components you need to complete a particular task. This makes unit testing testing considerably faster, but it also opens up flexible deployment opportunities, such as optimising the performance of certain production workloads by only loading particular subsets of your application.

Reviewer 2: ```suggestion
One of the most powerful features of Hanami 2 is your ability to partially boot the application, loading only the specific components you need to complete a particular task. This makes testing considerably faster, but it also opens up flexible deployment opportunities, such as optimising the performance of certain production workloads by only loading particular subsets of your application.
```

With this release, this flexibility has been extended to the application router and its Rack interface, which are now initialized lazily, allowing you to access the router and rack app even if your application has not fully booted. The rack app is now available as `.rack_app` on your application class, and can be accessed at any time after the initial "init" step:

```ruby
# Example config.ru

# Loads the Hanami app at config/application.rb and runs Hanami.init
#
# n.b. this does _not_ fully boot the application, so most components are not loaded
require "hanami/init"

run Hanami.rack_app
```

> Access to the Rack app in this way opens up the possibility of a slimeline deployment of a your application configured to serve only a small subset of its overall routes. In this case, only the minimal subset of components will be loaded to serve the given routes. This may be helpful for fine-grained deployment to resource-constrained targets like serverless functions.

Reviewer 2: ```suggestion
Access to the Rack app in this way opens up the possibility of a slimline deployment of a your application configured to serve only a small subset of its overall routes. In this case, only the minimal subset of components will be loaded to serve the given routes. This may be helpful for fine-grained deployment to resource-constrained targets like serverless functions.
```

## Route helpers via the view context

Last month we made the [route helpers available in actions](/blog/2021/12/07/announcing-hanami-200alpha4/), and now we’re making them available in views too, via our default [view context](https://dry-rb.org/gems/dry-view/0.7/context/). This means  you can access `routes` inside any template:

```slim
/ slices/main/web/templates/books/index.html.slim

- books.each do |book|
  p
    a href=routes.url(:book, id: book.id) = book.title
```

Given the context is exposed to view [parts](https://dry-rb.org/gems/dry-view/0.7/parts/) and [scopes](https://dry-rb.org/gems/dry-view/0.7/scopes/), you can access `routes` there as well.

## RSS support in our default MIME type list

You can now configure your Hanami actions to work with RSS formatted requests and responses via the new `rss` entry in our default MIME types list.

## What’s included?

Today we’re releasing the following gems:

> - hanami v2.0.0.alpha5

Reviewer: ```suggestion
- `hanami` `v2.0.0.alpha5`
```

> - hanami-controller v2.0.0.alpha5

Reviewer: ```suggestion
- `hanami-controller v2.0.0.alpha5`
```

> - hanami-view v2.0.0.alpha5

Reviewer: ```suggestion
- `hanami-view v2.0.0.alpha5`
```

      !component.identifier.start_with?("entities")
    }

    # Removing a default component dir
    config.source_dirs.component_dirs.delete("views")
  end
end
```

The `config.source_dirs.component_dirs` setting is provided by [dry-system’s](http://dry-rb.org/gems/dry-system/0.21/) own `Dry::System::Config::ComponentDirs`, so you can use this to configure any aspect of component loading as you require.

In addition to component dirs, you can also configure your own autoload paths, for source files you don’t want registered as components, but whose classes you still want to access inside your application. These directories are helpful for any classes that you will directly instantiate with their own data rather than dependencies, such as entities, structs, or any other kind of value class.

Out of the box, our `autoload_paths` is `["entities"]`. You can configure this as required:

```ruby
module MyApp
  class Application < Hanami::Application
    # Adding your own autoload paths
    config.source_dirs.autoload_paths << "structs"

    # Or providing a full replacement
    config.source_dirs.autoload_paths = ["structs"]
  end
end
```

## Lazy router and Rack app initialization

> One of the most powerful features of Hanami 2 is your ability to partially boot the application, loading only the specific components you need to complete a particular task. This makes testing testing considerably faster, but it also opens up flexible deployment opportunities, such as optimising the performance of certain production workloads by only loading particular subsets of your application.

Reviewer 2: ```suggestion
One of the most powerful features of Hanami 2 is your ability to partially boot the application, loading only the specific components you need to complete a particular task. This makes testing considerably faster, but it also opens up flexible deployment opportunities, such as optimising the performance of certain production workloads by only loading particular subsets of your application.
```

Author: @solnic lol, thank you!
